# About Me
I am a recent Carnegie Mellon University graduate with a Master's in computational biology, building on my experiences from my undergraduate education at Wheaton College. I am seeking to further develop my skillset and refine my interests so as to further pursue research in biology, mathematics, and computer science. 

My past research experience largely dealt with freshwater jellyfish, producing two publications and two conference presentations discussing the life cycle progression and the collection of these organisms. In continuing this research, I am currently working on collecting and analyzing imaging data with computer vision tools to further unravel specific transitions in the life of *Craspedacusta sowerbii* polyps, building on previous unpublished data.

I currently work as an adjunct instructor for Wheaton College teaching computer science, stemming from my wide range of experience as a TA. 

For a more detailed look at my work and experience, please consult [my CV](jonathanzhu_cv.pdf).

# My Work

## Publications

**J. A. Zhu**, N. C. Folino-Rorem. Effectiveness of Sampling Techniques for Collecting the Polyp Stage of the Invasive Freshwater Hydrozoan *Craspedacusta sowerbii*. *Biology* **2024**, *13*, 645. [https://doi.org/10.3390/biology13080645](https://doi.org/10.3390/biology13080645)
- Also see my repository with full data and data analysis code: [https://github.com/jonazhu/craspedacustasampling](https://github.com/jonazhu/craspedacustasampling)

K. Winata, **J. A. Zhu**, K. Hanselman, E. Zerbe, J. Langguth, N. C. Folino-Rorem, P. Cartwright. Life cycle transitions in the freshwater jellyfish *Craspedacusta sowerbii*. *Biology* **2024**, *13*, 1069. [https://doi.org/10.3390/biology13121069](https://doi.org/10.3390/biology13121069)

P. Isihara, **J. A. Zhu**. Clique enumeration in a math capstone course. (Accepted).

**J. A. Zhu**. Exploratory Data Analysis, in An Introduction to Python Jupyter Notebooks for College Math Teachers, pages 155-230. Open Textbook Library, [https://open.umn.edu/opentextbooks/textbooks/1697](https://open.umn.edu/opentextbooks/textbooks/1697).

**J. A. Zhu**, C. Wagner, Y. Li. Advanced Data Analysis, in An Introduction to Python Jupyter Notebooks for College Math Teachers, pages 1015-1115. Open Textbook Library, [https://open.umn.edu/opentextbooks/textbooks/1697](https://open.umn.edu/opentextbooks/textbooks/1697).

## Poster and Oral Presentations

**J. A. Zhu**, N. C. Folino-Rorem. “A novel sampling method for collecting the polyp stage of the invasive freshwater hydrozoan *Craspedacusta sowerbii* in aquatic habitats.” 23rd International Conference on Aquatic Invasive Species (ICAIS), Halifax, Nova Scotia, May 12-16, 2024.

A. Anil, L. Michal, X. L. Zhang, **J. A. Zhu**. “Enhancing Drug Discovery for Circadian Rhythm Regulation through Toxicity Classification of CRY1 Protein Molecules.” Carnegie Mellon University 2024 Computational Biology Project Poster Presentations, Pittsburgh, PA, May 2024.

K. Winata, **J. A. Zhu**, J. Langguth, C. G. Davis, M. Perry, N. C. Folino-Rorem, P. Cartwright. “Life Cycle Transitions in the Freshwater Jellyfish, *Craspedacusta sowerbii*.” Society for Integrative and Comparative Biology (SICB) Annual Meeting 2024, Seattle, WA, January 2-6, 2024.

**J. A. Zhu**, P. A. Jantsch. “A Decision Tree Model and Misclassification Metric to Predict Relative Risks of Property Tax Sales in Chicago’s South Side.” Wheaton College 2022 Sciences Research Poster Presentations, Wheaton, IL, October 2022. 

## Graduate Work Repositories
* Automation of Scientific Research: Designed and implemented stopping rules and criteria for advanced active learning algorithms in the context of cell video data embeddings. See the repository of our work [here](https://github.com/jonazhu/ACTIVibrio) and our technical writeup [here](reports/automation.pdf)
* Modeling Evolution: Ran simulations on various factors influencing extinction, including natural selection, invasive species, and habitat degradation. See the repository of our work [here](https://github.com/jonazhu/extinctionmodeling) and our technical writeup [here](reports/modevo.pdf)
* Biological Modeling and Simulation: Worked with a team in a two-part project to detect the positions of cells in microscope images and track these cells over time to extract insights. See the repository of our work [here](https://github.com/darin-tb/TrackLife) and our technical writeup [here](reports/modeling.pdf)
* Genomics and Epigenetics of the Brain: Conducted statistical analyses on open chromatin regions in the genomes of hundreds of mammals in relation to their neural traits, specifically hypselodonty (infinitely growing teeth). See the repository of our work [here](https://github.com/shashkat/Genomics-and-Epigenetics-of-the-Brain) and our technical writeup [here](reports/geb_main.pdf). We did further work looking at open chromatin regions in various cell types in relation to glioblastoma; see our technical writeup [here](reports/geb_secondary.pdf).
* Bioinformatics Data Science Integration Practicum: I worked with a team to produce a pipeline that reads protein sequences, introduces mutations specified by the user, and predicts the structural effects of these mutations. See the repository of our work [here](https://github.com/jonazhu/03713_bioinformatics_teamblue). 
* Machine Learning for Scientists: I worked with a team to predict the toxicity of various compounds using molecular features; the project includes several machine learning algorithms implemented without major library functions. See the repository of our work [here](https://github.com/jonazhu/02620_ml_group5) and our technical writeup [here](reports/ml.pdf).

### Other Graduate Work
* Neural Networks and Deep Learning in Science: I worked with a partner to develop a procedure for generation of chest x-ray images with various conditions to improve model training and make image datasets more balanced. See our technical writeup [here](reports/deeplearning.pdf). (Note: this project does not have a code repository associated with it as it was developed alongside my partner's research, and code will be made available upon publication.)
* Programming for Scientists: I made a pipeline using Golang and R that takes species occurrence data, cleans it, and predicts its potential spread. This resulted in the [pipeline itself](programming_project), [a writeup](programming_project/programming_report.pdf), and a [presentation](programming_project/programming_presentation.pdf).
* Practical Computing for Biologists: I made a [Python script](bact_genome) that reads a bacterial genome and produces an output file of all possible protein-coding sequences. The script can also translate these sequences and BLAST them against a database.

**Other relevant courses**: Neural Networks and Deep Learning in Science, Computational Molecular Biology and Genomics, Algorithms and Advanced Data Structures, Applied Cell and Molecular Biology, Data Science for Biological Scientists, Practical Computing for Biologists, Introduction to Statistical Genetics, Quantitative Genetics

## Undergraduate Work
For a detailed look at some key highlights of my undergraduate work, [click here](undergrad_work.md).